File: boca/J1_08/Department.java
import java.util.ArrayList;

public class Department {

    private String name;
    private Employee[] emps;

    public Department(String name, int num_emps) {
        this.name = name;
        this.emps = new Employee[num_emps];
    }

    public void addEmployee(Employee emp) {
        if(this.emps[this.emps.length - 1] != null) {
            Employee[] copy = this.emps;
            this.emps = new Employee[this.emps.length + 1];
            for(int i = 0; i < copy.length; i++) {
                this.emps[i] = copy[i];
            }
            this.emps[this.emps.length - 1] = emp;
        }

        else {
            for(int i = 0; i < this.emps.length; i++) {
                if(this.emps[i] == null) {
                    this.emps[i] = emp;
                    break;
                }
            }
        }
    }

    public Employee removeEmployee(int index) {
        Employee toRemove = this.emps[index];

        for(int i = index; i < this.emps.length - 1; i++) {
            this.emps[i] = this.emps[i + 1];
        }

        this.emps[this.emps.length - 1] = null;

        return toRemove;
    }

    public void moveEmployeeTo(int index, Department other) {
        Employee emp = this.removeEmployee(index);
        other.addEmployee(emp);
    }

    public String getName() {
        return this.name;
    }

    public double getSalary() {
        double sum;
        sum = new Double(0);
        for(int i = 0; i < this.emps.length; i++) {
            if(this.emps[i] != null)
                sum += this.emps[i].getSalary();
        }

        return sum;
    }

    public int getNumEmps() {
        return this.emps.length;
    }

}

File: boca/C1_02/Department.cpp
//
// Created by ulian on 21/06/19.
//

#include "Department.h"
#include "StringUtils.h"

using namespace std;

Department::Department(const string& name) {
    this->name.clear();
    this->name.append(name);
    this->emps.clear();
}

Department::~Department() {
    this->emps.clear();
}

void Department::addEmployee(Employee emp) {
    this->emps.insert(this->emps.end(), emp);
}

string Department::getName() {
    return this->name;
}

Employee* Department::popEmployee() {
    auto* emp = new Employee(this->emps.back());
    this->emps.pop_back();
    return emp;
}

double Department::getSalariesSum() {
    double sum = 0;

    for(Employee e : this->emps) {
       sum += e.getSalary();
    }

    return sum;
}

int Department::size() {
    return this->emps.size();
}

void Department::increaseSalaryPercentage(const double& percentage) {

    for(int i = 0; i < this->emps.size(); i++) {
        this->emps[i].changeSalary(this->emps[i].getSalary() + this->emps[i].getSalary()*percentage);
    }

}

Employee Department::getEmployee(const int& index) {
    Employee ret = this->emps[index];
    this->emps.erase(this->emps.begin() + index);
    return ret;
}